charmhelpers/contrib/openstack/context.py

```diff
@@ -2577,6 +2581,26 @@ def cpu_mask(self):
                 mask = mask | 1 << core
         return format(mask, '#04x')
 
+    def pmd_cpu_mask(self):
+        """Get hex formatted pmd CPU mask
+
+        The mask is based on config:pmd-cpu-set.
+        :returns: hex formatted CPU mask
+        :rtype: str
+        """
+        mask = 0
+        cpu_list = self._parse_cpu_list(config('pmd-cpu-set'))
+        if cpu_list:
+            for core in cpu_list:
+                mask = mask | 1 << core
+        else:
+            # The default mask is based on the next cpu of each NUMA node
+            # following the dpdk-lcore-mask in order to avoid overlap.
+            num_cores = config('dpdk-socket-cores')
+            for cores in self._numa_node_cores().values():
+                mask = mask | 1 << cores[num_cores]
```

Comment:
Would you mind enlightening me the reasoning behind this being the best default behavior? Any reason why we could/should not use the [Open vSwitch default behavior](https://github.com/openvswitch/ovs/blob/f686957c9667ae962fb8fc003be2a5482e380d75/vswitchd/vswitch.xml#L338-L339) in the event of this value not being set by the end user of the charm?

Reply:
I think I see the logic here - this ensures that the PMD and DPDK lcore don't overlap - the PMD is placed on the core after the allocation for the DPDK lcore

Reply:
> I think I see the logic here - this ensures that the PMD and DPDK lcore don't overlap - the PMD is placed on the core after the allocation for the DPDK lcore

I don't disagree with that, my problem with the current proposal is that if the user does not specify this config the charm does not need to populate the OVS config at all, [OVS will create a PMD thread per NUMA node and pin it to a available core all on its own.](https://github.com/openvswitch/ovs/blob/5db8aa39d9b7f6317f0d450e503de53865fa5593/vswitchd/vswitch.xml#L354-L355)

Since the original author has apparently disappeared and we want this resolved I guess we just need to fork the change.

Reply:
@fnordahl, I did not go away ;). In fact I had updated the initial [bug](https://bugs.launchpad.net/charm-ovn-chassis/+bug/1905416/comments/7) with additional information following this merged [PR](https://github.com/juju/charm-helpers/pull/538) which was following the same idea as this PR but was not compatible with my [layer-ovn PR](https://github.com/openstack-charmers/charm-layer-ovn/pull/41) which actually exposes the config-option to the ovn-chassis charm.

Reply:
Excellent. So the thing is that I would still like this change to land.

While adding a automatic get-me-a-number-of-pmd cores option is great, the reality is that if you're interesting in tuning these parameters, you most likely will have very specific needs. I.e. the system will most likely have a set of CPU cores isolated, and you would want PMD threads spawned on those specific cores.

So I would very much like this PR to land as well. If you have a spare moment to update the offending section, i.e. don't set a mask if the config is not populated, I think we should be good to go.

Reply:
Happy to make the change, though I now recall the reason behind this default behavior so will share it just in case:
In the [link](https://github.com/openvswitch/ovs/blob/5db8aa39d9b7f6317f0d450e503de53865fa5593/vswitchd/vswitch.xml#L354-L355) you mentioned it states:
> If not specified, one pmd thread will be created for each numa node
   and pinned to **_any_** available core on the numa node by default.

But I also came across this [link](https://developers.redhat.com/blog/2017/06/28/ovs-dpdk-parameters-dealing-with-multi-numa#dpdk_init) which mentions:
> OVS-DPDK performance typically scales well with additional cores, so it is common that users set pmd-cpu-mask but if it is not set, _**the default is that the lowest core on each NUMA node will be used for datapath packet processing**_.

As a result, if defaults are used for both dpdk-socket-cores and pmd-cpu-set, then you would end up with an overlap between dpdk-lcore-mask and pmd-cpu-mask (and the ovn-chassis charm would go into blocked state according to my layer-ovn PR).

Reply:
Thanks for the update, in that case we should stop the lcore mask from being configured into OVS unless it's actually set by the end user and require both to be configured together when set. We'll deal with that in charm code

Reply:
And to be clear, that means we want the code to handle a new `pmd-cpu-set` configuration option (i.e. this PR) to land, sans the default mask when config not set. I.e. if user does not provide config let OVS defaults (which may develop over time) run the show.

Reply:
Change has been made, I've also removed the previous implementation brought by 3b7a86420629cc45ced7f78fa3706d99f514448e to avoid confusion. I checked the ovn and ovs charms and I didn't see that config option being used yet so I assumed it was safe to remove.